### PR TITLE
Set up client API code generation

### PR DESCRIPTION
**Pull request by pamella**
## Description

We will use the [`openapi-ts`](https://heyapi.vercel.app/openapi-ts/get-started.html) tool to generate TypeScript client code for the frontend based on the backend schema generated by [`DRF-Spectacular`](https://drf-spectacular.readthedocs.io/en/latest/readme.html).

## Screenshots (if appropriate):

[Screencast from 14-05-2024 17:39:51.webm](https://github.com/vintasoftware/django-react-boilerplate/assets/22326737/56fbe929-ec2f-4bc4-9793-f7acf007f387)

## Steps to reproduce (if appropriate):

- The project should be running properly.
   - With docker
   - Without docker
   - Deployed (e.g.: Render)
- The API documentation pages should be working properly.

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [X] My code follows the code style of this project.
- [X] My change requires documentation updates.
- [X] I have updated the documentation accordingly.
- [X] My change requires dependencies updates.
- [X] I have updated the dependencies accordingly.

### CODE REVIEW

File: backend/common/views.py [hunk 1]
```diff
@@ -1,35 +1,34 @@
...
 
+from .serializers import MessageSerializer
+
 
 class IndexView(generic.TemplateView):
     template_name = "common/index.html"
 
 
 class RestViewSet(viewsets.ViewSet):
+    serializer_class = MessageSerializer
```
Comment (pamella) [inline backend/common/views.py:17]: I added the serializer so we can have a better usage example in the frontend.

File: frontend/js/api/index.ts
Comment (pamella) [inline frontend/js/api/index.ts:1]: Empty file needed so the GitHub build action doesn't break.

File: frontend/js/pages/Home.tsx [hunk 2]
```diff
@@ -1,20 +1,20 @@
 import { useState, useEffect } from "react";
 import Button from "react-bootstrap/Button";
-import { useDispatch, useSelector } from "react-redux";
 
 import DjangoImgSrc from "../../assets/images/django-logo-negative.png";
-import { AppDispatch, RootState } from "../store";
-import { fetchRestCheck } from "../store/rest_check";
```
Comment (pamella) [inline frontend/js/pages/Home.tsx:7]: @fjsj Let me know if you believe we should remove those files.

File: .eslintignore [hunk 3]
```diff
@@ -1,2 +1,3 @@
 frontend/bundles/
 frontend/webpack_bundles/
+frontend/js/api/
```
Comment (pamella) [inline .eslintignore:3]: Ignore auto-generated files by `@hey-api/openapi-ts`.

File: package.json [hunk 4]
```diff
@@ -37,14 +39,14 @@
...
-    "@swc/core": "^1.5.5",
+    "@swc/core": "^1.5.6",
     "@swc/jest": "^0.2.36",
     "@testing-library/jest-dom": "~6.4.5",
     "@testing-library/react": "~15.0.7",
     "@testing-library/user-event": "~14.5.2",
     "@types/cookie": "^0.6.0",
     "@types/jest": "^29.5.12",
-    "@types/node": "^20.12.11",
+    "@types/node": "^20.12.12",
```
Comment (pamella) [inline package.json:49]: Drive-by change.

**Comment by pamella** [comment]
Note: the `security/snyk` step is failing due to an `Uncontrolled resource consumption` issue on https://github.com/micromatch/braces. The issue was reported yesterday and there isn't a patch yet.

File: .github/workflows/shared-build/action.yml [hunk 5]
```diff
@@ -11,15 +11,15 @@ runs:
...
-      uses: actions/setup-python@v2
+      uses: actions/setup-python@v5
       with:
         python-version: "3.12"
     - name: Setup Node
-      uses: actions/setup-node@v2
+      uses: actions/setup-node@v4
       with:
-        node-version: "18.2"
+        node-version: "18.20"
```
Comment (fjsj) [inline .github/workflows/shared-build/action.yml:20]: We should upgrade to Node 20. Please tackle in another PR.

File: README.md [hunk 6]
```diff
@@ -204,19 +207,28 @@ Will run django tests using `--keepdb` and `--parallel`. You may pass a path to
...
 > - If you are not using Docker: `poetry run python manage.py spectacular --color --file schema.yml`
 
+We use the [`openapi-ts`](https://heyapi.vercel.app/openapi-ts/get-started.html) tool to generate TypeScript client code from the OpenAPI schema. The generated client code is used to interact with the API in a type-safe manner.
+
+> [!IMPORTANT]
+> Anytime the API schema is updated, the client code must be regenerated to reflect the changes. Failing to do so can lead to type errors in the client code.
+>
+> To update the client code, run:
+> - If you are using Docker: `make docker_frontend_update_api`
+> - If you are not using Docker: `npm run openapi-ts`
```
Comment (fjsj) [inline README.md:230]: If this happens on pre-commit, please explain here.

File: frontend/js/pages/Home.tsx (hunk 2 above)
Comment (fjsj) [reply frontend/js/pages/Home.tsx:7]: Yes, please do.

**Review by fjsj — CHANGES_REQUESTED** [review]
Overall LGTM, but can we add client generation to pre-commit?

**Comment by fjsj** [comment]
Nice @pamella , then all we need to merge here is passing snyk and explaining the pre-commit client generation on README.

File: frontend/js/pages/Home.tsx (hunk 2 above)
Comment (pamella) [reply frontend/js/pages/Home.tsx:7]: Done (commit 66dd971)

**Comment by pamella** [comment]
I've made the fixes. I guess we will just have to wait for `braces` to fix the vulnerability now. Hopefully, the will merge the in-progress PR soon :crossed_fingers:

File: .github/workflows/shared-build/action.yml (hunk 5 above)
Comment (pamella) [reply .github/workflows/shared-build/action.yml:20]: Here is the PR: #657

**Comment by pamella** [comment]
> I think we need to remove redux mentions from README. @pamella

Sorry, I missed it. Done (commit 37d8ba0)

**Review by fjsj — APPROVED** [review]
